src/Development/IDE/Core/Shake.hs

```diff
@@ -1170,3 +1181,51 @@ updatePositionMapping IdeState{shakeExtras = ShakeExtras{positionMapping}} Versi
         pure $! HMap.insert uri updatedMapping allMappings
   where
     shared_change = mkDelta changes
+
+startTelemetry :: ShakeExtras -> IO ()
+startTelemetry shakeExtras = do
+    IdeOptions{ optOTMemoryProfiling = (IdeOTMemoryProfiling otMemoryProfilingEnabled) } <- getIdeOptionsIO shakeExtras
+
+    when otMemoryProfilingEnabled $ do
+        let ShakeExtras { state=stateRef } = shakeExtras
+        instrumentFor <- getInstrumentCached <$> (newVar HMap.empty)
+
+        mapBytesInstrument <- mkValueObserver ("value map size_bytes")
+        mapCountInstrument <- mkValueObserver ("values map count")
+
+        _ <- regularly 50000 $ -- 100 times/s
+            withSpan_ "Measure length" $
+            readVar stateRef
+            >>= observe mapCountInstrument . length
+
+        _ <- regularly 0 $
+            withSpan_ "Measure Memory" $ do
+            performGC
+            values <- readVar stateRef
+            let groupedValues =
+                    HMap.toList values
+                    <&> (\((f, k), v) -> (k, [(f, v)]))
+                    & HMap.fromListWith (++)
+            valuesSize <- sequence $ HMap.mapWithKey (\k v -> withSpan ("Measure " <> (BS.pack $ show k)) $ \sp -> do
+                    { instrument <- instrumentFor k
+                    ; byteSize <- (sizeOf (undefined :: Word) *) <$> recursiveSizeNoGC v
```

Comment:
This call to `recursiveSizeNoGC` leads to a segfault if a GC arises while we are counting, which is quite easy to reproduce.

Given the way `recursiveSizeNoGC` works this is impossible to avoid (the act of measuring the size itself allocates memory) but it could be minimised by stopping the world while measuring.

Did you consider this and/or other workarounds @mpardalos ?

Reply:
Another, much cheaper way to minimise the segfaults in `recursiveSizeNoGC` is to use `System.Mem.Weak` to observe GC runs, and abort the count when this happens.

@mpardalos are you accepting PRs for heapsize?

Reply:
Yeah absolutely

Reply:
I have tested the `Weak` workaround and bingo, no more segfaults. So I'm going to:

* Send you a heapsize PR
* Ask you to upload a new heapsize version to Hackage 
* Commandeer this diff (I have made plenty of changes)
* Get this merged

Reply:
https://github.com/mpardalos/Heapsize/pull/1

Reply:
https://github.com/mpardalos/ghcide/pull/1

Reply:
I spoke too soon, segfaults can still arise if GC strikes in between the Weak check and the address dereferencing. It would be much safer if this code was written in C and accessed through an unsafe foreign import, because that guarantees that a GC cannot strike.

I also investigated a solution for the sharing problems, providing an Applicative interface so that it's possible to do multiple heapsize measurements with the same set of seen pointers. Again, the GC gets in the way - making this approach not viable sadly. One would need to rewrite the whole thing in C.